### PR TITLE
Fix for #290. Work correctly if HW offloading is absent

### PR DESCRIPTION
**Pull request by gshimansky**
Applications now have a way to determine HW offloading capabilities
and control their use by NFF-Go framework.

### CODE REVIEW

File: low/low.h [hunk 1]
```diff
@@ -615,3 +615,15 @@ static int KNI_config_promiscusity(uint16_t port_id, uint8_t to_on) {
...
+bool check_hwtxchecksum_capability(uint16_t port_id) {
+    uint64_t flags = DEV_TX_OFFLOAD_IPV4_CKSUM |
+        DEV_TX_OFFLOAD_UDP_CKSUM |
+        DEV_TX_OFFLOAD_TCP_CKSUM;
+
+	struct rte_eth_dev_info dev_info;
+	memset(&dev_info, 0, sizeof(dev_info));
+	rte_eth_dev_info_get(port_id, &dev_info);
+
+    return (dev_info.tx_offload_capa & flags) == flags;
```
Comment (ifilippov) [inline low/low.h]: Tabs like above

File: low/low.h [hunk 2]
```diff
@@ -615,3 +615,15 @@ static int KNI_config_promiscusity(uint16_t port_id, uint8_t to_on) {
 	fprintf(stderr, "DEBUG: KNI: Promiscusity mode of port %d %s\n", port_id, to_on ? "on" : "off");
 	return 0;
 }
+
+bool check_hwtxchecksum_capability(uint16_t port_id) {
```
Comment (ifilippov) [inline low/low.h]: You should return false also if port index exceeded existing ports number

File: flow/flow.go [hunk 3]
```diff
@@ -319,6 +319,35 @@ func addSegment(in *low.Ring, first *Func) *processSegment {
 	return segment
 }
 
+type HWCapability int
+
+const (
+	HWTXChecksumCapability HWCapability = iota
```
Comment (ifilippov) [inline flow/flow.go]: What the idea about more complex code? What do you expect to add in this constant?

Comment (gshimansky) [reply flow/flow.go]: You can refer to rte_ethdev.h for a list possible offloading capabilities. We may support some of them in the future.
```
/**
 * RX offload capabilities of a device.
 */
#define DEV_RX_OFFLOAD_VLAN_STRIP  0x00000001
#define DEV_RX_OFFLOAD_IPV4_CKSUM  0x00000002
#define DEV_RX_OFFLOAD_UDP_CKSUM   0x00000004
#define DEV_RX_OFFLOAD_TCP_CKSUM   0x00000008
#define DEV_RX_OFFLOAD_TCP_LRO     0x00000010
#define DEV_RX_OFFLOAD_QINQ_STRIP  0x00000020
#define DEV_RX_OFFLOAD_OUTER_IPV4_CKSUM 0x00000040
#define DEV_RX_OFFLOAD_MACSEC_STRIP     0x00000080
#define DEV_RX_OFFLOAD_HEADER_SPLIT	0x00000100
#define DEV_RX_OFFLOAD_VLAN_FILTER	0x00000200
#define DEV_RX_OFFLOAD_VLAN_EXTEND	0x00000400
#define DEV_RX_OFFLOAD_JUMBO_FRAME	0x00000800
#define DEV_RX_OFFLOAD_CRC_STRIP	0x00001000
#define DEV_RX_OFFLOAD_SCATTER		0x00002000
#define DEV_RX_OFFLOAD_TIMESTAMP	0x00004000
#define DEV_RX_OFFLOAD_SECURITY         0x00008000
#define DEV_RX_OFFLOAD_CHECKSUM (DEV_RX_OFFLOAD_IPV4_CKSUM | \
				 DEV_RX_OFFLOAD_UDP_CKSUM | \
				 DEV_RX_OFFLOAD_TCP_CKSUM)
#define DEV_RX_OFFLOAD_VLAN (DEV_RX_OFFLOAD_VLAN_STRIP | \
			     DEV_RX_OFFLOAD_VLAN_FILTER | \
			     DEV_RX_OFFLOAD_VLAN_EXTEND)

/*
 * If new Rx offload capabilities are defined, they also must be
 * mentioned in rte_rx_offload_names in rte_ethdev.c file.
 */

/**
 * TX offload capabilities of a device.
 */
#define DEV_TX_OFFLOAD_VLAN_INSERT 0x00000001
#define DEV_TX_OFFLOAD_IPV4_CKSUM  0x00000002
#define DEV_TX_OFFLOAD_UDP_CKSUM   0x00000004
#define DEV_TX_OFFLOAD_TCP_CKSUM   0x00000008
#define DEV_TX_OFFLOAD_SCTP_CKSUM  0x00000010
#define DEV_TX_OFFLOAD_TCP_TSO     0x00000020
#define DEV_TX_OFFLOAD_UDP_TSO     0x00000040
#define DEV_TX_OFFLOAD_OUTER_IPV4_CKSUM 0x00000080 /**< Used for tunneling packet. */
#define DEV_TX_OFFLOAD_QINQ_INSERT 0x00000100
#define DEV_TX_OFFLOAD_VXLAN_TNL_TSO    0x00000200    /**< Used for tunneling packet. */
#define DEV_TX_OFFLOAD_GRE_TNL_TSO      0x00000400    /**< Used for tunneling packet. */
#define DEV_TX_OFFLOAD_IPIP_TNL_TSO     0x00000800    /**< Used for tunneling packet. */
#define DEV_TX_OFFLOAD_GENEVE_TNL_TSO   0x00001000    /**< Used for tunneling packet. */
#define DEV_TX_OFFLOAD_MACSEC_INSERT    0x00002000
#define DEV_TX_OFFLOAD_MT_LOCKFREE      0x00004000
/**< Multiple threads can invoke rte_eth_tx_burst() concurrently on the same
 * tx queue without SW lock.
 */
#define DEV_TX_OFFLOAD_MULTI_SEGS	0x00008000
/**< Device supports multi segment send. */
#define DEV_TX_OFFLOAD_MBUF_FAST_FREE	0x00010000
/**< Device supports optimization for fast release of mbufs.
 *   When set application must guarantee that per-queue all mbufs comes from
 *   the same mempool and has refcnt = 1.
 */
#define DEV_TX_OFFLOAD_SECURITY         0x00020000
```

File: low/low.h (hunk 1 above)
Comment (gshimansky) [reply low/low.h]: Ok

File: low/low.h (hunk 2 above)
Comment (gshimansky) [reply low/low.h]: Ok

File: flow/flow.go [hunk 4]
```diff
@@ -319,6 +319,35 @@ func addSegment(in *low.Ring, first *Func) *processSegment {
...
+)
+
+// CheckHWCapability return true if hardware offloading capability
+// present in all ports. Otherwise it returns false.
+func CheckHWCapability(capa HWCapability, ports []uint16) bool {
+	for p := range ports {
+		switch capa {
+		case HWTXChecksumCapability:
+			if !low.CheckHWTXChecksumCapability(ports[p]) {
```
Comment (ifilippov) [inline flow/flow.go]: As I understand capa can have only one value, if it is right - you can just return low level function output without any checking.

Comment (gshimansky) [reply flow/flow.go]: Other values are supposed to be added in the future, so it makes sense to write the handling code from the beginning.

**Comment by ifilippov** [comment]
Approved

**Pull request merged**